week7/answers.sql

```diff
@@ -1,3 +1,280 @@
+    IF equipped_item_damage > 0 THEN
+        SET target_character_health = target_character_health - equipped_item_damage;
+
+        -- Update health in the character stats table
+        UPDATE character_stats
+        SET health = target_character_health
+        WHERE character_id = target_character_id;
+
+        -- If the character's health reaches 0 or below, they die
+        IF target_character_health <= 0 THEN
+            -- Delete the character and all related records
+            DELETE FROM character_stats WHERE character_id = target_character_id;
+            DELETE FROM equipped WHERE character_id = target_character_id;
+            DELETE FROM inventory WHERE character_id = target_character_id;
+            DELETE FROM team_members WHERE character_id = target_character_id;
+        END IF;
+    END IF;
+END $$
+
+CREATE PROCEDURE equip(in _inventory_id INT)
+BEGIN
+    -- Declare variables to store character_id and item_id
+    DECLARE _character_id INT;
+    DECLARE _item_id INT;
+
+    -- Get the character_id and item_id from the inventory table
+    SELECT character_id, item_id
+    INTO _character_id, _item_id
+    FROM inventory
+    WHERE inventory_id = _inventory_id;
+
+    -- Remove the item from the inventory table
+    DELETE FROM inventory WHERE inventory_id = _inventory_id;
+
+    -- Add the item to the equipped table
+    INSERT INTO equipped (character_id, item_id)
+    VALUES (_character_id, _item_id);
+END $$
+
+CREATE PROCEDURE unequip(in _equipped_id INT)
+BEGIN
+    -- Declare variables to store character_id and item_id
+    DECLARE _character_id INT;
+    DECLARE _item_id INT;
+
+    -- Get the character_id and item_id from the equipped table
+    SELECT character_id, item_id
+    INTO _character_id, _item_id
+    FROM equipped
+    WHERE equipped_id = _equipped_id
+    LIMIT 1;
+
+    -- Remove the item from the equipped table
+    DELETE FROM equipped WHERE equipped_id = _equipped_id;
+
+    -- Add the item back to the inventory table
+    INSERT INTO inventory (character_id, item_id)
+    VALUES (_character_id, _item_id);
+END $$
+  
+CREATE PROCEDURE set_winners(IN team_id INT)
+BEGIN
+    DECLARE done INT DEFAULT 0;
+    DECLARE character_id INT;
+    DECLARE cursor_winners CURSOR FOR
+        SELECT tm.character_id
+        FROM team_members tm
+        JOIN character_stats cs ON tm.character_id = cs.character_id
+        WHERE tm.team_id = team_id AND cs.health > 0;
+
+ -- Declare a handler to handle the end of the cursor
+    DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = 1;
+
+    -- Empty the winners table
+    DELETE FROM winners;
+
+    -- Open the cursor
+    OPEN cursor_winners;
+
+    -- Loop through each result
+    read_loop: LOOP
+        FETCH cursor_winners INTO character_id;
+        IF done THEN
+            LEAVE read_loop;
+        END IF;
+
+        -- Insert each winner into the winners table
+        INSERT INTO winners (character_id, name)
```

Comment:
This query can be vastly simplified by getting the name with the CURSOR, which also guarantees that you cannot get a DUPLICATE KEY. Nothing in this assignment requires a nested SELECT statement.